internal/characters/xilonen/skill.go

```diff
@@ -0,0 +1,203 @@
+	samplerShredKey  = "xilonen-e-shred"
+	activeSamplerKey = "xilonen-samplers-activated"
+)
+
+func init() {
+	skillFrames = frames.InitAbilSlice(65)
+	skillFrames[action.ActionAttack] = 19
+	skillFrames[action.ActionBurst] = 20
+	skillFrames[action.ActionDash] = 15
+	skillFrames[action.ActionJump] = 15
+	skillFrames[action.ActionSwap] = 19
+	skillFrames[action.ActionWalk] = 20
+}
+
+func (c *char) reduceNightsoulPoints(val float64) {
+	if c.StatusIsActive(c6key) {
+		return
+	}
+
+	c.nightsoulState.ConsumePoints(val * c.nightsoulConsumptionMul())
+	if c.nightsoulState.Points() <= 0.0001 {
+		c.exitNightsoul()
+	}
+}
+
+func (c *char) enterNightsoul() {
+	c.nightsoulState.EnterBlessing(45)
+	c.nightsoulSrc = c.Core.F
+	c.QueueCharTask(c.nightsoulPointReduceFunc(c.nightsoulSrc), 12)
+	c.NormalHitNum = rollerHitNum
+	c.NormalCounter = 0
+	c.samplersActivated = false
+
+	duration := int(9 * 60 * c.nightsoulDurationMul())
+	c.setNightsoulExitTimer(duration)
+
+	// Don't queue the task if C2 or higher
+	if c.Base.Cons < 2 && c.shredElements[attributes.Geo] {
+		c.activeGeoSampler(c.nightsoulSrc)()
+	}
+}
+
+func (c *char) exitNightsoul() {
+	if !c.nightsoulState.HasBlessing() {
+		return
+	}
+	c.nightsoulState.ExitBlessing()
+	c.nightsoulSrc = -1
+	c.exitStateSrc = -1
+	c.SetCDWithDelay(action.ActionSkill, 7*60, 0)
+	c.NormalHitNum = normalHitNum
+	c.NormalCounter = 0
+}
+
+func (c *char) nightsoulPointReduceFunc(src int) func() {
+	return func() {
+		if c.nightsoulSrc != src {
+			return
+		}
+		if !c.nightsoulState.HasBlessing() {
+			return
+		}
+
+		c.reduceNightsoulPoints(1)
+		// reduce 1 point per 12f, which is 5 per second
+		c.QueueCharTask(c.nightsoulPointReduceFunc(src), 12)
+	}
+}
+
+func (c *char) applySamplerShred(ele attributes.Element, enemies []combat.Enemy) {
+	for _, e := range enemies {
+		e.AddResistMod(combat.ResistMod{
+			Base:  modifier.NewBaseWithHitlag(fmt.Sprintf("%v-%v", samplerShredKey, ele.String()), 60),
+			Ele:   ele,
+			Value: -skillShred[c.TalentLvlSkill()],
+		})
+	}
+}
+
+func (c *char) activeGeoSampler(src int) func() {
+	return func() {
+		if c.Base.Cons < 2 {
+			if c.nightsoulSrc != src || !c.nightsoulState.HasBlessing() {
+				return
+			}
+		}
+		enemies := c.Core.Combat.EnemiesWithinArea(combat.NewCircleHitOnTarget(c.Core.Combat.Player(), nil, 10), nil)
+		c.applySamplerShred(attributes.Geo, enemies)
+		c.QueueCharTask(c.activeGeoSampler(src), samplerInterval)
+	}
+}
+
+func (c *char) activeSamplers(src int) func() {
+	return func() {
+		if c.sampleSrc != src {
+			return
+		}
+		if !c.StatusIsActive(activeSamplerKey) {
+			return
+		}
+
+		enemies := c.Core.Combat.EnemiesWithinArea(combat.NewCircleHitOnTarget(c.Core.Combat.Player(), nil, 10), nil)
+		for ele := range c.shredElements {
+			// skip geo when C2 or above since it's always active
+			if ele == attributes.Geo {
+				continue
+			}
+			c.applySamplerShred(ele, enemies)
+		}
+		c.QueueCharTask(c.activeSamplers(src), samplerInterval)
```

Comment:
what about getting an active character in the task instead of using the `OnCharacterSwap` event?
```go
func (c *char) activeSamplers(src int) func() {
	return func() {
		if c.sampleSrc != src {
			return
		}
		if !c.StatusIsActive(activeSamplerKey) {
			return
		}

		ch := c.Core.Player.ActiveChar()
		enemies := c.Core.Combat.EnemiesWithinArea(combat.NewCircleHitOnTarget(c.Core.Combat.Player(), nil, 10), nil)
		for ele := range c.shredElements {
			// skip geo when C2 or above since it's always active
			if ele == attributes.Geo {
				continue
			}
			c.applySamplerShred(ele, enemies)
		}
		ch.QueueCharTask(c.activeSamplers(src), samplerInterval)
	}
}
```

Reply:
the `3s` tick is reset on swap, i think what @Charlie-Zheng had was correct expect forgetting to set `c.sampleSrc` on swap.